appcelerator/titanium_mobile
# feat(android): add callback support to Ti.Platform.openURL()

**JIRA:**
https://jira.appcelerator.org/browse/TIMOB-27862

**Note:**
The following mocha suite PR will unit test this via app custom URL scheme.
https://github.com/appcelerator/titanium-mobile-mocha-suite/pull/250

**Test:**
1. Build and run the below code on Android.
2. Tap on the "OpenURL(url)" button.
3. Verify it launches a web browser showing Google's website.
4. Verify the log shows the below:
`@@@ openURL() returned: true`
5. Navigate back to the Titanium app.
6. Tap on the "OpenURL(url, callback)" button.
7. Verify it launches a web browser showing Google's website.
8. Verify the log shows the below. Event must be logged last with success set to `true`:
`@@@ openURL() returned: true`
`@@@ openURL() event: {"code":0,"success":true}`
9. Navigate back to the Titanium app.
10. Tap on the "OpenURL(url, options, callback)" button.
11. Verify it launches a web browser showing Google's website.
12. Verify the log shows the below. Event must be logged last with success set to `true`:
`@@@ openURL() returned: true`
`@@@ openURL() event: {"code":0,"success":true}`
13. Navigate back to the Titanium app.
14. Tap on the "OpenURL(bad-url)" button.
15. Verify that nothing launches. _(This is good. It has a bad URL.)_
16. Verify the log shows the below. Returned value must be `false`.
`@@@ openURL() returned: false`
17. Tap on the "OpenURL(bad-url)" button.
18. Verify that nothing launches. _(This is good. It has a bad URL.)_
19. Verify the log shows the below. Event must be logged last and success set to `false`:
`[INFO]  @@@ openURL() returned: false`
`[INFO]  @@@ openURL() event: {"code":1,"success":false}`

```javascript
var url = "https://www.google.com";
var badUrl = "idontexist://";

var window = Ti.UI.createWindow({
	layout: "vertical",
});
var open1Button = Ti.UI.createButton({
	title: "openURL(url)",
	top: Ti.App.iOS ? "40dp" : "20dp",
});
open1Button.addEventListener("click", function() {
	var result = Ti.Platform.openURL(url);
	Ti.API.info("@@@ openURL() returned: " + result);
});
window.add(open1Button);
var open2Button = Ti.UI.createButton({
	title: "openURL(url, callback)",
	top: "20dp",
});
open2Button.addEventListener("click", function() {
	var result = Ti.Platform.openURL(url, function(e) {
		Ti.API.info("@@@ openURL() event: " + JSON.stringify(e));
	});
	Ti.API.info("@@@ openURL() returned: " + result);
});
window.add(open2Button);
var open3Button = Ti.UI.createButton({
	title: "openURL(url, options, callback)",
	top: "20dp",
});
open3Button.addEventListener("click", function() {
	var openOptions = {
		UIApplicationOpenURLOptionsOpenInPlaceKey: true,
	};
	var result = Ti.Platform.openURL(url, openOptions, function(e) {
		Ti.API.info("@@@ openURL() event: " + JSON.stringify(e));
	});
	Ti.API.info("@@@ openURL() returned: " + result);
});
window.add(open3Button);
var open4Button = Ti.UI.createButton({
	title: "openURL(bad-url)",
	top: "20dp",
});
open4Button.addEventListener("click", function() {
	var result = Ti.Platform.openURL(badUrl);
	Ti.API.info("@@@ openURL() returned: " + result);
});
window.add(open4Button);
var open5Button = Ti.UI.createButton({
	title: "openURL(bad-url, callback)",
	top: "20dp",
});
open5Button.addEventListener("click", function() {
	var result = Ti.Platform.openURL(badUrl, function(e) {
		Ti.API.info("@@@ openURL() event: " + JSON.stringify(e));
	});
	Ti.API.info("@@@ openURL() returned: " + result);
});
window.add(open5Button);
window.open();
```
